File: C++/Knapsack.cpp
#include <iostream>
#include <vector>

using namespace std;

// Structure to represent an item with weight and value
struct Item {
    int weight;
    int value;
};

// Function to solve the 0/1 Knapsack Problem using dynamic programming
int knapsackDP(const vector<Item>& items, int capacity) {
    int n = items.size();
    // Create a DP table to store the maximum value for different item counts and capacities
    vector<vector<int>> dp(n + 1, vector<int>(capacity + 1, 0));

    for (int i = 1; i <= n; ++i) {
        for (int w = 0; w <= capacity; ++w) {
            // If the current item's weight can fit in the knapsack
            if (items[i - 1].weight <= w) {
                // Choose the maximum value between two options:
                // 1. Not including the current item
                // 2. Including the current item and adding its value
                dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - items[i - 1].weight] + items[i - 1].value);
            } else {
                // If the item's weight exceeds the current capacity, skip it
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    // The final value in the DP table represents the maximum value achievable
    return dp[n][capacity];
}

int main() {
    int capacity;
    cout << "Enter the knapsack capacity: ";
    cin >> capacity;

    // Define a list of items with their weights and values
    vector<Item> items = {
        {2, 12},
        {1, 10},
        {3, 20},
        {2, 15}
    };

    // Call the knapsackDP function to find the maximum value
    int maxValue = knapsackDP(items, capacity);

    // Display the result
    cout << "The maximum value that can be obtained is: " << maxValue << endl;

    return 0;
}

File: JavaScript/knapsack.js
function knapsack(values, weights, capacity) {
    // Get the number of items
    const n = values.length;

    // Build a 2D array to store results for subproblems
    const dp = new Array(n + 1).fill(null).map(() => new Array(capacity + 1).fill(0));

    // Loop through each item
    for (let i = 1; i <= n; i++) {
        // Loop through each possible capacity
        for (let w = 1; w <= capacity; w++) {
            // Check if the weight of the current item can fit in the current capacity
            if (weights[i - 1] <= w) {
                // Calculate the maximum value when including and excluding the current item
                dp[i][w] = Math.max(dp[i - 1][w], dp[i - 1][w - weights[i - 1]] + values[i - 1]);
            } else {
                // If the current item's weight exceeds the capacity, we can't include it
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    //  Variables to track the selected items
    let totalValue = dp[n][capacity];  // Total value of selected items
    let totalWeight = capacity;        // Remaining capacity
    let selectedItems = [];           // List of selected item indices

    // Backtracking to find the selected items
    for (let i = n; i > 0 && totalValue > 0; i--) {
        if (totalValue !== dp[i - 1][totalWeight]) {
            // If the value changes, the current item is included
            selectedItems.push(i - 1);  // Adjust for 0-based indexing
            totalValue -= values[i - 1];
            totalWeight -= weights[i - 1];
        }
    }
    return {
        totalValue: dp[n][capacity],         // Total value of the selected items
        totalWeight: capacity - totalWeight, // Total weight used
        selectedItems: selectedItems.reverse() // List of selected item indices
    };
}

// Example usage:
let values = [10, 20, 30];
let weights = [1, 2, 3];
let capacity = 5;
let result = knapsack(values, weights, capacity);
console.log(result);
